tests/cuda_custom_call/cuda_custom_call_test.py

```diff
@@ -0,0 +1,227 @@
+      operands=[a, b],
+      backend_config=dict(n=ir.IntegerAttr.get(u64, n)),
+      operand_layouts=[layout, layout],  # a, b
+      result_layouts=[layout, layout],  # c, b_plus_1
+  ).results
+
+
+# construct a new JAX primitive
+foo_fwd_p = jax.core.Primitive(JAX_PRIMITIVE_FWD)
+# register the abstract evaluation rule for the forward primitive
+foo_fwd_p.def_abstract_eval(_foo_fwd_abstract_eval)
+foo_fwd_p.multiple_results = True
+mlir.register_lowering(foo_fwd_p, _foo_fwd_lowering, platform=JAX_PLATFORM)
+
+#-----------------------------------------------------------------------------#
+#                              Backward pass                                  #
+#-----------------------------------------------------------------------------#
+
+# register the XLA FFI binding pointer with XLA
+xla_client.register_custom_call_target(name=XLA_CUSTOM_CALL_TARGET_BWD,
+                                       fn=ffi.pycapsule(library.FooBwd),
+                                       platform=XLA_PLATFORM,
+                                       api_version=XLA_CUSTOM_CALL_API_VERSION)
+
+
+def _foo_bwd_abstract_eval(c_grad, a, b_plus_1):
+  assert c_grad.shape == a.shape
+  assert a.shape == b_plus_1.shape
+  assert c_grad.dtype == a.dtype
+  assert a.dtype == b_plus_1.dtype
+
+  shaped_array = jax.core.ShapedArray(a.shape, a.dtype)
+  return (
+      shaped_array,  # a_grad
+      shaped_array,  # b_grad
+  )
+
+
+def _foo_bwd_lowering(ctx, c_grad, a, b_plus_1):
+  typ = ir.RankedTensorType(a.type)
+  shape = typ.shape
+  u64 = ir.IntegerType.get_unsigned(64)
+  n = np.prod(shape)
+  # default XLA layout that just means row major
+  layout = range(len(shape) - 1, -1, -1)
+  return custom_call(
+      XLA_CUSTOM_CALL_TARGET_BWD,
+      api_version=STABLEHLO_CUSTOM_CALL_API_VERSION,
+      result_types=[typ, typ],  # a_grad, b_grad
+      operands=[c_grad, a, b_plus_1],
+      backend_config=dict(n=ir.IntegerAttr.get(u64, n)),
+      operand_layouts=[layout, layout, layout],  # c_grad, a, b_plus_1
+      result_layouts=[layout, layout],  # a_grad, b_grad
+  ).results
+
+
+# construct a new JAX primitive
+foo_bwd_p = jax.core.Primitive(JAX_PRIMITIVE_BWD)
+# register the abstract evaluation rule for the backward primitive
+foo_bwd_p.def_abstract_eval(_foo_bwd_abstract_eval)
+foo_bwd_p.multiple_results = True
+mlir.register_lowering(foo_bwd_p, _foo_bwd_lowering, platform=JAX_PLATFORM)
+
+#-----------------------------------------------------------------------------#
+#                              User facing API                                #
+#-----------------------------------------------------------------------------#
+
+def foo_fwd(a, b):
+  c, b_plus_1 = foo_fwd_p.bind(a, b)
+  return c, (a, b_plus_1)
+
+
+def foo_bwd(res, c_grad):
+  a, b_plus_1 = res
+  return foo_bwd_p.bind(c_grad, a, b_plus_1)
+
+
+@jax.custom_vjp
+def foo(a, b):
+  c, _ = foo_fwd(a, b)
+  return c
+
+
+foo.defvjp(foo_fwd, foo_bwd)
```

Comment:
I'm not convinced that this test case needs to test `custom_vjp`, since that is tested elsewhere. I recognize that it's useful to demonstrate how one might use `custom_vjp` for exposing derivatives, but doesn't that live in the docs rather than the tests?

Removing the `backward` op would also dramatically simplify this file!